### PR TITLE
is_hdd update for windows and linux

### PR DESCRIPTION
**Pull request by moneromooo-monero**
Add is_hdd_sysfs(), is_hdd_win_ioctl(), fl2lv(), lv2pv(),
    win_device_has_seek_penalny()
Add tests test_a.is_hdd*
Modify is_hdd()
Added bcrypt, sodium

Contributed by crCr62U0_ who does not have nor want a github account.

### CODE REVIEW

File: src/common/util.cpp [hunk 1]
```diff
@@ -74,6 +75,11 @@ using namespace epee;
 #include <boost/asio.hpp>
 #include <openssl/sha.h>
 
+#ifdef WIN32
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: I think we had problems with WIN32 before, and _WIN32 was the "best" one to use.

File: src/common/util.cpp [hunk 2]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
+    if(f == nullptr)
+    {
+      attr_path = prefix + "/../queue/rotational";
+      f = fopen(attr_path.c_str(), "r");
+      if(f == nullptr)
+      {
+        return false;
+      }
+    }
+    uint16_t val = 0xdead;
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: Probably best to make it a unsigned short if using %hu below.

File: src/common/util.cpp [hunk 3]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
+        return true;
+      }
+    }
+    return false;
+  }
 
-    std::string sys_path = "/sys/block/" + device + "/queue/rotational";
-    FILE *f = fopen(sys_path.c_str(), "r");
-    if (!f)
+  bool win_device_has_seek_penalny(const char *pv, bool &result)
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: penalty

File: src/common/util.cpp [hunk 4]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
+        &q,
+        sizeof(q),
+        &r,
+        sizeof(r),
+        &r_size,
+        nullptr
+        );
+      if(success and r_size == sizeof(r))
+      {
+        result = r.IncursSeekPenalty;
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: That path seems to leak h.

File: src/common/util.cpp [hunk 5]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
-    return 0;
+    }
+    std::vector<std::string> pvs;
+    bool pv_success = lv2pv(lv.c_str(), pvs);
+    if(not pv_success)
+    {
+      return false;
+    }
+    bool a_result = 0;
+    bool a_success = 0;
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: false

File: src/common/util.cpp [hunk 6]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
+  }
 #endif
+  bool is_hdd(const char *path, bool &result)
+  {
+    #if defined(_WIN32) and (_WIN32_WINNT >= 0x0601)
+    return is_hdd_win_ioctl(path, result);
+    #elif defined(__GLIBC__)
+    return is_hdd_sysfs(path, result);
+    #else
+    return 0;
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: false

File: src/common/util.cpp [hunk 7]
```diff
@@ -733,63 +739,225 @@ std::string get_nix_version_display_string()
...
+  bool is_hdd(const char *path)
+  {
+    bool result;
+    if(is_hdd(path, result))
+    {
+        return result;
+    }
+    else
+    {
+        return 0;
```
Comment (moneromooo-monero) [inline src/common/util.cpp]: false

File: tests/unit_tests/CMakeLists.txt [hunk 8]
```diff
@@ -72,6 +72,7 @@ set(unit_tests_sources
   ringct.cpp
   output_selection.cpp
   vercmp.cpp
+  test_a.cpp
```
Comment (moneromooo-monero) [inline tests/unit_tests/CMakeLists.txt]: Rename to something sane

**Comment by p8p** [comment]
I don't have write access to push updates.

**Comment by moneromooo-monero** [comment]
If you are the author, then you can open a new bug for it. If you are not, then I suggest you ping the author in #monero-dev (nick in the first post) about it. I PRed because the author did not want to make a github account.

**Comment by moneromooo-monero** [comment]
+duplicate

**Pull request closed without merging**